### PR TITLE
Introduces monaco-editor

### PR DESCRIPTION
**Pull request by KaiVolland**
## Description

This replaces the editor used for the `CodeEditor` component.
[Code-Mirror](https://codemirror.net/) is replaced with [Monaco-Editor](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-configure-json-defaults) wrapped by [`@monaco-editor/react`](https://www.npmjs.com/package/@monaco-editor/react).

monaco-screenshot

The monaco-editor comes with tons of extra features and optional plugins. For example, we can use the new geostyler-style JSON schema (https://github.com/geostyler/geostyler-style/pull/154) to validate, document and autocomplete the entered geostyler-style.

The old code-mirror editor is not removed completly but moved to the `Legacy` folder in components.

As this PR relies on https://github.com/geostyler/geostyler-style/pull/154 (and a corresponding release) it is currently in draft mode but can be reviewed allready.

## Pull request type

- [ ] Bugfix
- [x] Feature
- [x] Dependency updates
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)
- [ ] I am unsure (we'll look into it together)

## Do you introduce a breaking change?

- [x] Yes
- [ ] No
- [ ] I am unsure (no worries, we'll find out)

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the [BSD 2-Clause License](https://github.com/geostyler/geostyler/blob/master/)
- [x] I have followed the [guidelines for contributing](https://github.com/geostyler/geostyler/blob/master/CONTRIBUTING.md)
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/geostyler/geostyler/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally)
- [ ] I'm lost; why do I have to check so many boxes? Please help!

### CODE REVIEW

**Comment by KaiVolland** [comment]
Sadly i had to remove a lot of tests. They all try to access methods of the component which are not exposed anymore as i transformed the class component into a function component.

File: src/Component/CodeEditor/CodeEditor.tsx [hunk 1]
```diff
@@ -93,204 +86,131 @@ export interface CodeEditorProps extends Partial<CodeEditorDefaultProps> {
...
-    });
-  }
-
-  updateValueFromStyle = (style: GsStyle) => {
-    this.valueFromStyleInput(style)
-      .then((parsedStyle: string) => {
-        this.setState({
-          value: parsedStyle
-        });
+const MODELPATH = 'geostlyer.json';  // associate with our model
```
Comment (dnlkoch) [inline src/Component/CodeEditor/CodeEditor.tsx:90]:
```suggestion
const MODELPATH = 'geostyler.json';  // associate with our model
```

File: src/Component/CodeEditor/CodeEditor.tsx [hunk 2]
```diff
@@ -93,204 +86,131 @@ export interface CodeEditorProps extends Partial<CodeEditorDefaultProps> {
...
+  delay = 500,
+  locale = en_US.GsCodeEditor,
+  onStyleChange,
+  parsers = [],
+  showCopyButton = false,
+  showSaveButton = false,
+  style
+}) => {
+
+  let editTimeout: any;
```
Comment (dnlkoch) [inline src/Component/CodeEditor/CodeEditor.tsx:106]:
```suggestion
  let editTimeout: number;
```

File: src/Component/CodeEditor/CodeEditor.tsx [hunk 3]
```diff
@@ -333,81 +247,58 @@ export class CodeEditor extends React.Component<CodeEditorProps, CodeEditorState
...
-          }
-        </div>
+  const parserHasUnitSelect = isSldParser && activeParser && (activeParser as SldStyleParser).sldVersion !== '1.0.0';
+
+  return (
+    <div className="gs-code-editor">
+      <div className="gs-code-editor-toolbar" >
+        {locale.formatSelectLabel}: <Select
+          className="gs-code-editor-format-select"
+          style={{ width: 300 }}
```
Comment (dnlkoch) [inline src/Component/CodeEditor/CodeEditor.tsx:250]: Not related to this PR in particular, but style should be moved to the style definition.

File: src/Component/CodeEditor/CodeEditor.tsx [hunk 4]
```diff
@@ -333,81 +247,58 @@ export class CodeEditor extends React.Component<CodeEditorProps, CodeEditorState
...
+        path={isSldParser ? undefined : MODELPATH}
+        language={isSldParser ? 'xml' : 'json'}
+        onChange={handleOnChange}
+      />
+      <div className="gs-code-editor-errormessage">
+        {invalidMessage}
+      </div>
+      <div className="gs-code-editor-bottombar">
+        {
+          !showCopyButton ? null :
```
Comment (dnlkoch) [inline src/Component/CodeEditor/CodeEditor.tsx:273]:
```suggestion
          showCopyButton &&
```

File: src/Component/CodeEditor/CodeEditor.tsx [hunk 5]
```diff
@@ -333,81 +247,58 @@ export class CodeEditor extends React.Component<CodeEditorProps, CodeEditorState
...
+            <Button
+              className="gs-code-editor-copy-button"
+              type="primary"
+              onClick={onCopyButtonClick}
+            >
+              {locale.copyButtonLabel}
+            </Button>
+        }
+        {
+          !showSaveButton ? null :
```
Comment (dnlkoch) [inline src/Component/CodeEditor/CodeEditor.tsx:283]:
```suggestion
          showSaveButton &&
```

**Review by dnlkoch — APPROVED** [review]

**Comment by KaiVolland** [comment]
Thanks for the review @dnlkoch

File: src/hook/UsePrevious.tsx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { useEffect, useRef } from 'react';
+
+export const usePrevious = <T extends unknown>(value: T): T | undefined => {
```
Comment (jansule) [inline src/hook/UsePrevious.tsx:3]: Minor: Some explanatory comments would be nice.

File: src/hook/UsePrevious.tsx [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { useEffect, useRef } from 'react';
```
Comment (jansule) [inline src/hook/UsePrevious.tsx:1]: Could you please add a license header?

**Review by jansule — CHANGES_REQUESTED** [review]

**Review by jansule — APPROVED** [review]
Nice work @KaiVolland! This is ready to merge now.

**Pull request merged**